stumpy/ostinato.py

```diff
@@ -485,6 +486,7 @@ def ostinatoed(client, Ts, m, normalize=True, p=2.0, Ts_subseq_isconstant=None):
     if not isinstance(Ts, list):  # pragma: no cover
         raise ValueError(f"`Ts` is of type `{type(Ts)}` but a `list` is expected")
 
+    Ts = [T.copy() for T in Ts]
```

Comment:
Since we are copying arrays here, should we do `copy=False` in the for-loop below?

```
for i, T in enumerate(Ts):
        Ts[i], M_Ts[i], Σ_Ts[i], Ts_subseq_isconstant[i] = core.preprocess(
            T, m, copy=False, T_subseq_isconstant=Ts_subseq_isconstant[i]
        )
```

I tested it, and it works.

----

Btw, the default value for `copy` is `True` in the `core.preprocess`. I am curious to know why `Ts[i], ... = core.preprocess(T, ....)`  did not work, and resulted in overwriting the array

Reply:
> Btw, the default value for copy is True in the core.preprocess. I am curious to know why Ts[i], ... = core.preprocess(T, ....) did not work, and resulted in overwriting the array

Hopefully, I am understanding your question correctly but I don't think it has to do with the `numpy` `array` level but instead with the `python` `list` level. Let's say we remove `numpy` out of the equation:

```
def some_func(Ts):
    for i in range(len(Ts)):
        Ts[i] = i + 10

if __name__ == '__main__':
    Ts = [1, 2, 3, 4]
    some_func(Ts)
    print(Ts)
    # 10 11 12 13
```

Since `python` is always "pass-by-reference", a `python` `list` is NOT a copy when passed into a function and, instead, you have direct access to the original list, which you are then overwriting the contents of. In the case of `core.preprocess`, we are filling/overwriting `Ts[i]` with a copy of a preprocessed array. So, to state it more clearly, I am guessing that the `core.preprocess` makes a copy of the array `T == Ts[i]`, processes it, and replaces the `i`th element in `Ts` with this new preprocessed data. I think we should leave `copy=True` (default) here.

Reply:
> should we do copy=False in the for-loop below?

Why? Is it purely to avoid a double `copy`? I think we should leave copy=True (default) here.

Reply:
> > Btw, the default value for copy is True in the core.preprocess. I am curious to know why Ts[i], ... = core.preprocess(T, ....) did not work, and resulted in overwriting the array
> 
> Hopefully, I am understanding your question correctly but I don't think it has to do with the `numpy` `array` level but instead with the `python` `list` level. Let's say we remove `numpy` out of the equation:
> 
> ```
> def some_func(Ts):
>     for i in range(len(Ts)):
>         Ts[i] = i + 10
> 
> if __name__ == '__main__':
>     Ts = [1, 2, 3, 4]
>     some_func(Ts)
>     print(Ts)
>     # 10 11 12 13
> ```
> 
> Since `python` is always "pass-by-reference", a `python` `list` is NOT a copy when passed into a function and, instead, you have direct access to the original list, which you are then overwriting the contents of.

You understood it correctly indeed. The example and its following explanation were very helpful! 👍  Thanks!

Reply:
@seanlaw 

> > should we do copy=False in the for-loop below?
> 
> Why? Is it purely to avoid a double copy? 

Yes. 

> I think we should leave copy=True (default) here. Otherwise, I would make a note above it just to say something like:
> `# We've already made a copy above`

I might be wrong but I feel your concern is coming from our future selves. I think you want to be on the safer side and leave it as `copy=True` (particularly if its contribution to the running time is negligible compared to the other parts of the code). This makes sense! I can see that this is a little price we agree to pay to keep ourselves in a better position in the long term.   So, I will go with your initial suggestion: 
> I think we should leave copy=True (default) here.

Reply:
If anything, you can add a comment to say "setting `copy=False` is also fine here if performance becomes an issue"

Reply:
@NimaSarajpoor I think I still prefer this:

```
# Avoid overwriting `Ts` during preprocessing
Ts_copy = [None] * len(Ts)
.
.
.
for i, T in enumerate(Ts):
        Ts_copy[i], M_Ts[i], Σ_Ts[i], Ts_subseq_isconstant[i] = core.preprocess(
            T, m, copy=True, T_subseq_isconstant=Ts_subseq_isconstant[i]
        )
```

In fact, if you go back to the original issue #983, my initial instinct was to propose something like:

```
Ts_copy = []
for T in Ts:
    Ts_copy.append(T.copy())
``` 

I am liking this better because:
1. We are being explicit with the copy
2. We gain the (tiny) performance benefit of not doing a double copy while having zero doubt that `Ts_copy` is a "copy" and therefore `copy=False` is fine.

Can we go with using the `Ts_copy` approach, please?

Reply:
@seanlaw 
I can understand the  "being the explicit" part. It makes sense. I will make the changes. Thanks for the feedback!  (The only reason I went with `Ts = [T.copy() for T in Ts]` was to avoid making changes in other parts of the code. But, I can see that was not a very good reason here)